sql: make UDF more robust with db/schema renaming

chengxiong-ruan opened the pull request:
There are 5 commits:
1. disallow cross-db references in udf. Also a side effect to serialize user defined type names in udf body as type OIDs.
2. check udf dependencies before renaming a db.
3. check dependencies before renaming a schema.
4. deserialize sequence OID for `SHOW CREATE FUNCTION`.
5. deserialize type OID for `SHOW CREATE FUNCTION`.

Release justification: necessary and low risk bug fixes to make UDF more robust with schema/db renaming.

chengxiong-ruan committed:
sql: disallow cross-db references in udf

This commit makes `CREATE FUNCTION` return an error when objects from
a different database are used in the function definition. A side effort
is made to rewrite user-defined type names into type OIDs, this is
actually where cross-db type references are blocked when type are resolved
within a db context.

Release note: None.
Release justification: fixing a bug where udfs can use relations and
types from other databases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87203)
<!-- Reviewable:end -->

chengxiong-ruan commented on pkg/sql/logictest/testdata/logic_test/udf:
```diff
@@ -1211,6 +1211,66 @@ CREATE FUNCTION public.f_test_cor(IN a INT8, IN b INT8)
+    CALLED ON NULL INPUT
+    LANGUAGE SQL
+    AS $$
+    SELECT b'@':::@100115;
```
oops.. this is bad.. need to reverse the rewrite when doing `show create`...
will fix this

chengxiong-ruan replied:
this is fixed partially in the last commit. also created #87475 to track the issue.

ajwerner reviewed (commented):
It seems to me like the rewrite is at least as important. Can you add a bunch of testing related to renaming the schema, database, and type name of a referenced type and make sure that the udf does not break? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-:-NAplNKo3N9-lQe9VLyG:b8gg4pk)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

chengxiong-ruan committed:
sql: check function dependencies when renaming databases

Previously, we only check table dependencies when renaming
a database. With UDF introduced, we need to handle function
depencies as well.

Release note: None.
Release justification: necessary bug fix with UDF introduced.

chengxiong-ruan committed:
sql: check dependencies properly when renaming a schema

Previously, we don't check table depedencies when renaming
a schema. This is problematic because views and UDFs can be
broken if it references a table from the schema being renamed.
This is pr add dependency checks to schema renaming by reusing
the logic of renaming databases.

Release note (sql change): previously, a schema can be renamed even when a
child table is referenced by a view or UDF. The renaming breaks the view
and UDF since tables are referenced by names in views and UDFs at the
moment. This commit adds check to disallow renaming a schema if there is
table references.
Release justification: low risk bug fix.

chengxiong-ruan committed:
sql: deserialize sequence name for `SHOW CREATE FUNCTION`

Previously, `SHOW CREATE FUNCTION` returns a function body
with sequence OIDs. This commit deserializes sequence OID
into qualified names.

Release note: None.
Release justification: low risk enhancement.

chengxiong-ruan committed:
sql: deserialize user defined type OIDs in UDF body

Release note: None.
Release justification: low risk display enhancement.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 3 files at r1, 2 of 2 files at r2, 6 of 6 files at r3, 5 of 5 files at r4, 3 of 3 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-:-NBNK3DiEWGXhqmtRb2V:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/rename_database.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNIruBASFNtldRWgVq:-NBNIruCE8TEE_A-WGgi:bc3xy7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/rename_database.go#L158)):*
> ```Go
> 			// Since we now only reference sequences with IDs, it's always safe to
> 			// rename a db or schema containing the sequence.
> 			if tbDesc.IsSequence() {
> ```

you're going to hate me -- this isn't quite true. You can use a sequence like a table, and it won't be rewritten. We may need to file yet another issue for that and we may want to find a way to disallow it. 



_Code quote:_
```Go
			// Since we now only reference sequences with IDs, it's always safe to
			// rename a db or schema containing the sequence.
			if tbDesc.IsSequence() {
```
___
*[`pkg/sql/schema_resolver.go` line 241 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNJ6Y3AacOH2puHJ6j:-NBNJ6Y3AacOH2puHJ6k:bo2r7g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/schema_resolver.go#L241)):*
> ```Go
> 	}
> 	scDesc, err := sr.descCollection.GetImmutableSchemaByID(ctx, sr.txn, fnDesc.GetParentSchemaID(),
> 		tree.SchemaLookupFlags{
> ```

why is this one not `Required`? In practice, it does not matter, but it still seems weird to me. My preference would be that you define a flags variable and cast it


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87203)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[`pkg/sql/rename_database.go` line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNIruBASFNtldRWgVq:-NBO1yzB1BWbvUdDKse5:b-h66qfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/rename_database.go#L158)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

you're going to hate me -- this isn't quite true. You can use a sequence like a table, and it won't be rewritten. We may need to file yet another issue for that and we may want to find a way to disallow it. 
</blockquote></details>

leaving it as is and filed #87509 to track this. note that this is also broken on master.
Also added test cases for this bad behavior. Thanks.
___
*[`pkg/sql/schema_resolver.go` line 241 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87203#-NBNJ6Y3AacOH2puHJ6j:-NBO2Mol1ECAiydXKFrL:b-j94wx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/90b962a72e4c0c748dddedf3f2875473b44bdd8f/pkg/sql/schema_resolver.go#L241)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why is this one not `Required`? In practice, it does not matter, but it still seems weird to me. My preference would be that you define a flags variable and cast it
</blockquote></details>

Good point, fixed, thanks!


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
TFTR!
bors r+

ajwerner commented:
blathers backport 22.2
